File: python/packages/kagent-adk/src/kagent/adk/tools/share_tools.py
"""Share link tools for agents to manage session sharing."""

from __future__ import annotations

import logging
import os
from typing import Any, Dict

import httpx
from google.adk.tools import BaseTool, ToolContext
from google.genai import types

logger = logging.getLogger("kagent_adk." + __name__)

_KAGENT_UI_URL = os.getenv("KAGENT_UI_URL", "").rstrip("/")


def _parse_app_name(app_name: str) -> tuple[str, str]:
    """Parse a Python-identifier app_name back to (namespace, name)."""
    parts = app_name.split("__NS__", 1)
    if len(parts) != 2:
        return ("", app_name.replace("_", "-"))
    return (parts[0].replace("_", "-"), parts[1].replace("_", "-"))


def _share_url(token: str, session_id: str, app_name: str) -> str:
    """Return the share URL for the current session.

    When KAGENT_UI_URL is set, returns the full URL. Otherwise returns the
    relative path ``/agents/<namespace>/<name>/chat/<session_id>?share=<token>``.
    """
    namespace, name = _parse_app_name(app_name)
    path = f"/agents/{namespace}/{name}/chat/{session_id}?share={token}"
    return f"{_KAGENT_UI_URL}{path}" if _KAGENT_UI_URL else path


class CreateShareLinkTool(BaseTool):
    """Create a share link for the current session.

    The link allows any authenticated user to view (and optionally interact with) the session.
    """

    def __init__(self, client: httpx.AsyncClient) -> None:
        super().__init__(
            name="create_share_link",
            description=(
                "Creates a share link for the current chat session. "
                "Returns a URL any authenticated user can open to view this session. "
                "The link is read-only by default (visitors cannot send messages). "
                "Set read_only=false to allow visitors to interact. "
                "Each call creates a new token; existing tokens remain valid."
            ),
        )
        self._client = client

    def _get_declaration(self) -> types.FunctionDeclaration:
        return types.FunctionDeclaration(
            name=self.name,
            description=self.description,
            parameters=types.Schema(
                type=types.Type.OBJECT,
                properties={
                    "read_only": types.Schema(
                        type=types.Type.BOOLEAN,
                        description="When true, the shared link will be read-only (visitors cannot send messages). Defaults to true.",
                    ),
                },
            ),
        )

    async def run_async(self, *, args: Dict[str, Any], tool_context: ToolContext) -> str:
        session_id = tool_context.session.id
        if not session_id or not session_id.strip():
            return "Error: session ID is empty — cannot create share link."
        app_name = tool_context.session.app_name
        read_only = bool(args.get("read_only", True))
        try:
            response = await self._client.post(
                f"/api/sessions/{session_id}/shares",
                json={"read_only": read_only},
            )
            if response.status_code == 201:
                data = response.json().get("data", {})
                token = data.get("token", "")
                suffix = " (read-only)" if read_only else ""
                return f"Share link created{suffix}: {_share_url(token, session_id, app_name)}"
            return f"Failed to create share link: HTTP {response.status_code}: {response.text}"
        except httpx.TimeoutException as e:
            logger.error("Timeout creating share link: %s", e)
            return "Error creating share link: request timed out"
        except httpx.RequestError as e:
            logger.error("Request error creating share link: %s", e)
            return f"Error creating share link: {e}"
        except Exception as e:
            logger.error("Error creating share link: %s", e)
            return f"Error creating share link: {e}"


class ListShareLinksTool(BaseTool):
    """List existing share links for the current session."""

    def __init__(self, client: httpx.AsyncClient) -> None:
        super().__init__(
            name="list_share_links",
            description=(
                "Lists all active share links for the current session. "
                "Returns each share token and creation time. "
                "Use this to find a token before calling delete_share_link."
            ),
        )
        self._client = client

    def _get_declaration(self) -> types.FunctionDeclaration:
        return types.FunctionDeclaration(
            name=self.name,
            description=self.description,
            parameters=types.Schema(type=types.Type.OBJECT, properties={}),
        )

    async def run_async(self, *, args: Dict[str, Any], tool_context: ToolContext) -> str:
        session_id = tool_context.session.id
        if not session_id or not session_id.strip():
            return "Error: session ID is empty — cannot list share links."
        try:
            response = await self._client.get(
                f"/api/sessions/{session_id}/shares",
            )
            if response.status_code == 200:
                shares = response.json().get("data", [])
                if not shares:
                    return "No active share links for this session."
                lines = [
                    f"- token: {s.get('token', '<unknown>')}, created_at: {s.get('created_at', 'unknown')}"
                    for s in shares
                ]
                return "Active share links:\n" + "\n".join(lines)
            return f"Failed to list share links: HTTP {response.status_code}: {response.text}"
        except httpx.TimeoutException as e:
            logger.error("Timeout listing share links: %s", e)
            return "Error listing share links: request timed out"
        except httpx.RequestError as e:
            logger.error("Request error listing share links: %s", e)
            return f"Error listing share links: {e}"
        except Exception as e:
            logger.error("Error listing share links: %s", e)
            return f"Error listing share links: {e}"


class DeleteShareLinkTool(BaseTool):
    """Delete a share link for the current session, revoking visitor access."""

    def __init__(self, client: httpx.AsyncClient) -> None:
        super().__init__(
            name="delete_share_link",
            description=(
                "Deletes a share link by token, immediately revoking access for anyone using it. "
                "Use list_share_links first to find the token you want to revoke."
            ),
        )
        self._client = client

    def _get_declaration(self) -> types.FunctionDeclaration:
        return types.FunctionDeclaration(
            name=self.name,
            description=self.description,
            parameters=types.Schema(
                type=types.Type.OBJECT,
                properties={
                    "token": types.Schema(
                        type=types.Type.STRING,
                        description="The share token to revoke.",
                    ),
                },
                required=["token"],
            ),
        )

    async def run_async(self, *, args: Dict[str, Any], tool_context: ToolContext) -> str:
        session_id = tool_context.session.id
        if not session_id or not session_id.strip():
            return "Error: session ID is empty — cannot delete share link."
        token = args.get("token", "").strip()
        if not token:
            return "Error: token is required."
        try:
            response = await self._client.delete(
                f"/api/sessions/{session_id}/shares/{token}",
            )
            if response.status_code == 200:
                return f"Share link {token!r} revoked successfully."
            return f"Failed to delete share link: HTTP {response.status_code}: {response.text}"
        except httpx.TimeoutException as e:
            logger.error("Timeout deleting share link: %s", e)
            return "Error deleting share link: request timed out"
        except httpx.RequestError as e:
            logger.error("Request error deleting share link: %s", e)
            return f"Error deleting share link: {e}"
        except Exception as e:
            logger.error("Error deleting share link: %s", e)
            return f"Error deleting share link: {e}"

File: go/adk/pkg/tools/share_tools.go
package tools

import (
	"context"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/kagent-dev/kagent/go/adk/pkg/controllerclient"
	apiv1alpha1 "github.com/kagent-dev/kagent/go/api/gen/kagent/api/v1alpha1"
	"google.golang.org/adk/v2/agent"
	"google.golang.org/adk/v2/tool"
	"google.golang.org/adk/v2/tool/functiontool"
)

// shareClient holds the dependencies for share link tools, captured at construction time.
type shareClient struct {
	controllerClient *controllerclient.Client
	uiURL            string // KAGENT_UI_URL, used to build full share URLs
	appName          string
}

// parseAppName converts a Python-identifier app_name back to (namespace, name).
// Format: "namespace__NS__agent_name" with hyphens encoded as underscores.
func parseAppName(appName string) (namespace, name string) {
	parts := strings.SplitN(appName, "__NS__", 2)
	if len(parts) != 2 {
		return "", strings.ReplaceAll(appName, "_", "-")
	}
	return strings.ReplaceAll(parts[0], "_", "-"), strings.ReplaceAll(parts[1], "_", "-")
}

// shareURL returns the share URL for a session token.
// With uiURL set it returns a full absolute URL; otherwise a relative path.
func (c *shareClient) shareURL(token, sessionID string) string {
	ns, name := parseAppName(c.appName)
	path := fmt.Sprintf("/agents/%s/%s/chat/%s?share=%s", ns, name, sessionID, token)
	if c.uiURL != "" {
		return c.uiURL + path
	}
	return path
}

func (c *shareClient) callContext(ctx context.Context, userID string) (context.Context, context.CancelFunc) {
	return c.controllerClient.CallContext(ctx, userID)
}

func shareToMap(share *apiv1alpha1.SessionShare) map[string]any {
	createdAt := time.Time{}
	if share.GetCreatedAt() != nil {
		createdAt = share.GetCreatedAt().AsTime()
	}
	return map[string]any{
		"id":         share.GetId(),
		"token":      share.GetToken(),
		"session_id": share.GetSessionId(),
		"user_id":    share.GetUserId(),
		"read_only":  share.GetReadOnly(),
		"created_at": createdAt.Format(time.RFC3339Nano),
	}
}

func (c *shareClient) createShare(ctx context.Context, userID, sessionID string, readOnly *bool) (*apiv1alpha1.SessionShare, error) {
	callContext, cancel := c.callContext(ctx, userID)
	defer cancel()
	response, err := c.controllerClient.SessionService().CreateSessionShare(callContext, &apiv1alpha1.CreateSessionShareRequest{
		SessionId: sessionID,
		ReadOnly:  readOnly,
	})
	if err != nil {
		return nil, err
	}
	if response.GetShare() == nil {
		return nil, fmt.Errorf("response did not include a share")
	}
	return response.GetShare(), nil
}

func (c *shareClient) listShares(ctx context.Context, userID, sessionID string) ([]any, error) {
	callContext, cancel := c.callContext(ctx, userID)
	defer cancel()
	response, err := c.controllerClient.SessionService().ListSessionShares(callContext, &apiv1alpha1.ListSessionSharesRequest{SessionId: sessionID})
	if err != nil {
		return nil, err
	}
	shares := make([]any, 0, len(response.GetShares()))
	for _, share := range response.GetShares() {
		shares = append(shares, shareToMap(share))
	}
	return shares, nil
}

func (c *shareClient) deleteShare(ctx context.Context, userID, sessionID, token string) error {
	callContext, cancel := c.callContext(ctx, userID)
	defer cancel()
	_, err := c.controllerClient.SessionService().DeleteSessionShare(callContext, &apiv1alpha1.DeleteSessionShareRequest{
		SessionId: sessionID,
		Token:     token,
	})
	return err
}

type createShareInput struct {
	// ReadOnly controls whether the shared link allows visitors to send messages.
	// When nil (not provided by the model), the server defaults to true (read-only).
	ReadOnly *bool `json:"read_only,omitempty"`
}

// NewCreateShareLinkTool creates a tool that generates a share token for the current session.
func NewCreateShareLinkTool(controllerClient *controllerclient.Client, appName string) (tool.Tool, error) {
	if controllerClient == nil {
		return nil, fmt.Errorf("controller client is required")
	}
	c := &shareClient{
		controllerClient: controllerClient,
		uiURL:            strings.TrimRight(os.Getenv("KAGENT_UI_URL"), "/"),
		appName:          appName,
	}
	return functiontool.New(functiontool.Config{
		Name: "create_share_link",
		Description: "Creates a share link for the current chat session. " +
			"Returns a URL any authenticated user can open to view this session. " +
			"The link is read-only by default (visitors cannot send messages). " +
			"Set read_only=false to allow visitors to interact. " +
			"Each call creates a new token; existing tokens remain valid.",
	}, func(ctx agent.Context, in createShareInput) (map[string]any, error) {
		sessionID := ctx.SessionID()
		if sessionID == "" {
			return nil, fmt.Errorf("create_share_link: no session ID in context")
		}
		share, err := c.createShare(ctx, ctx.UserID(), sessionID, in.ReadOnly)
		if err != nil {
			return nil, fmt.Errorf("create_share_link: %w", err)
		}
		return map[string]any{
			"url":       c.shareURL(share.GetToken(), sessionID),
			"read_only": share.GetReadOnly(),
		}, nil
	})
}

// NewListShareLinksTool creates a tool that lists active share tokens for the current session.
func NewListShareLinksTool(controllerClient *controllerclient.Client, appName string) (tool.Tool, error) {
	if controllerClient == nil {
		return nil, fmt.Errorf("controller client is required")
	}
	c := &shareClient{
		controllerClient: controllerClient,
		uiURL:            strings.TrimRight(os.Getenv("KAGENT_UI_URL"), "/"),
		appName:          appName,
	}
	return functiontool.New(functiontool.Config{
		Name: "list_share_links",
		Description: "Lists all active share links for the current session. " +
			"Returns each share token and creation time. " +
			"Use this to find a token before calling delete_share_link.",
	}, func(ctx agent.Context, _ struct{}) (map[string]any, error) {
		sessionID := ctx.SessionID()
		if sessionID == "" {
			return nil, fmt.Errorf("list_share_links: no session ID in context")
		}
		shares, err := c.listShares(ctx, ctx.UserID(), sessionID)
		if err != nil {
			return nil, fmt.Errorf("list_share_links: %w", err)
		}
		return map[string]any{"shares": shares}, nil
	})
}

type deleteShareInput struct {
	Token string `json:"token"`
}

// NewDeleteShareLinkTool creates a tool that revokes a specific share token for the current session.
func NewDeleteShareLinkTool(controllerClient *controllerclient.Client, appName string) (tool.Tool, error) {
	if controllerClient == nil {
		return nil, fmt.Errorf("controller client is required")
	}
	c := &shareClient{
		controllerClient: controllerClient,
		uiURL:            strings.TrimRight(os.Getenv("KAGENT_UI_URL"), "/"),
		appName:          appName,
	}
	return functiontool.New(functiontool.Config{
		Name: "delete_share_link",
		Description: "Deletes a share link by token, immediately revoking access for anyone using it. " +
			"Use list_share_links first to find the token you want to revoke.",
	}, func(ctx agent.Context, in deleteShareInput) (map[string]any, error) {
		if in.Token == "" {
			return nil, fmt.Errorf("delete_share_link: token is required")
		}
		sessionID := ctx.SessionID()
		if sessionID == "" {
			return nil, fmt.Errorf("delete_share_link: no session ID in context")
		}
		if err := c.deleteShare(ctx, ctx.UserID(), sessionID, in.Token); err != nil {
			return nil, fmt.Errorf("delete_share_link: %w", err)
		}
		return map[string]any{"status": "revoked", "token": in.Token}, nil
	})
}
